src/Utilities/EditorConfigFinder.cs

```diff
@@ -1,45 +1,41 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
 
+using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.IO;
 
 namespace Microsoft.CodeAnalysis.Tools.Utilities
 {
     internal static class EditorConfigFinder
     {
-        public static ImmutableArray<string> GetEditorConfigPaths(string path)
+        public static ImmutableArray<string> GetEditorConfigPaths(ImmutableArray<string> paths)
         {
-            // If the path is to a file then remove the file name and process the
-            // folder path.
-            var startPath = Directory.Exists(path)
-                ? path
-                : Path.GetDirectoryName(path);
-
-            if (!Directory.Exists(startPath))
+            if (paths == default)
             {
                 return ImmutableArray<string>.Empty;
             }
 
             var editorConfigPaths = ImmutableArray.CreateBuilder<string>(16);
+            var visited = new List<string>();
 
-            var directory = new DirectoryInfo(path);
-
-            // Find .editorconfig files contained unders the folder path.
-            var files = directory.GetFiles(".editorconfig", SearchOption.AllDirectories);
-            for (var index = 0; index < files.Length; index++)
+            foreach (var path in paths)
             {
-                editorConfigPaths.Add(files[index].FullName);
-            }
+                var directoryName = Path.GetDirectoryName(path);
+                if (visited.Contains(directoryName)) continue;
 
-            // Walk from the folder path up to the drive root addings .editorconfig files.
-            while (directory.Parent != null)
-            {
-                directory = directory.Parent;
+                visited.Add(directoryName);
+                var directory = new DirectoryInfo(directoryName);
 
-                files = directory.GetFiles(".editorconfig", SearchOption.TopDirectoryOnly);
-                if (files.Length == 1)
+                // Walk from the folder path up to the drive root addings .editorconfig files.
+                while (directory.Parent != null)
                 {
-                    editorConfigPaths.Add(files[0].FullName);
+                    directory = directory.Parent;
+
+                    var files = directory.GetFiles(".editorconfig", SearchOption.TopDirectoryOnly);
```

Comment:
We should check whether the directory has been visited and add the directory to visited within this loop, since many source directories will share the same branches.